Editable taxonomy definitions

alexkim205 opened the pull request:
## Changes

Make new definition panels editable.

https://user-images.githubusercontent.com/13460330/152470946-06856021-cab6-4557-b28b-7abf6fc895ed.mov

This PR persists event and property definition to url in events/properties table page.

Todo before this PR is ready:
- [x] tests for new logic
- [x] when a user drags mouse outside of input box, hover card closes
- [x] disable hotkeys when interacting with hover card
- [x] cancel doesn't work perfectly. Split out state into local state and saved state.

## How did you test this code?

Logic tests

alexkim205 committed:
definitions now editable

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
fix error

mariusandra commented:
Very cool! I noticed from the video that when you changed the event's verified status, it changed the icon, but still said "verified event" as the title.

clarkus commented:
Regarding your outstanding items

> when a user drags mouse outside of input box, hover card closes

I think once the editing mode is triggered, the taxonomic editing card, the event picker, etc (all the way down the event chain) should persist open until that user clicks close or save on the modal. My concern is that a low-effort mouse action like moving your cursor cancels out of editing mid-edit. If we do automatically close the card before changes are saved, we should prompt the user to notify that content will be lost.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
still trying to figure out disabling interaction

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
less aggressive click and hotkey handlers

alexkim205 committed:
add tests and overlay styling

alexkim205 commented on ee/migrations/0008_null_verified_allowed.py:
```diff
@@ -0,0 +1,18 @@
+        ("ee", "0007_dashboard_permissions"),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name="enterpriseeventdefinition",
+            name="verified",
+            field=models.BooleanField(blank=True, default=False, null=True),
+        ),
+    ]
```
Allow null values for `verified` property so that api calls are backwards compatible (i.e., event definitions with undefined verified property)

alexkim205 committed:
console remove

alexkim205 committed:
owner dropdown

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
owner changing works

alexkim205 commented on ee/migrations/0008_null_def_properties_allowed.py:
```diff
@@ -0,0 +1,23 @@
+        ("ee", "0007_dashboard_permissions"),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name="enterpriseeventdefinition",
+            name="description",
+            field=models.CharField(blank=True, max_length=400, null=True),
+        ),
+        migrations.AlterField(
+            model_name="enterpriseeventdefinition",
+            name="verified",
+            field=models.BooleanField(blank=True, default=False, null=True),
```
Allow null values for verified and description property so that api calls are backwards compatible (i.e., event definitions with undefined properties)

alexkim205 committed:
hide owner for posthog props

alexkim205 committed:
add default description

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
remove irrelevant test

alexkim205 commented on frontend/src/lib/components/DefinitionPopup/definitionPopupLogic.ts:
```diff
@@ -1,21 +1,43 @@
+        setDefinition: (item: Partial<TaxonomicDefinitionTypes>) => ({ item }),
+        setLocalDefinition: (item: Partial<TaxonomicDefinitionTypes>) => ({ item }),
         setPopupState: (state: DefinitionPopupState) => ({ state }),
+        setNewTag: (tag: string) => ({ tag }),
+        deleteTag: (tag: string) => ({ tag }),
```
Note this api will change once https://github.com/PostHog/posthog/pull/8567 is merged in. This PR will probably be merged in after that one.

rcmarron commented on ee/models/event_definition.py:
```diff
@@ -6,11 +6,11 @@
     updated_at: models.DateTimeField = models.DateTimeField(auto_now=True)
     updated_by = models.ForeignKey("posthog.User", null=True, on_delete=models.SET_NULL, blank=True)
-    verified: models.BooleanField = models.BooleanField(default=False, null=False, blank=True)
+    verified: models.BooleanField = models.BooleanField(default=False, null=True, blank=True)
```
I think the null=true here makes the `verified` field have 3 options: Null, False and True. We probably want to keep it to just the 2 options (False and True)

rcmarron reviewed (dismissed):
Looks great! A few thoughts:

* Should the owner be visible in the pop-over? Feels a bit odd to add an owner inline, but not be able to see it in the view mode. If we don't think it's valuable, maybe we just remove it from the pop-up?
* I think it can be a bit jarring to click 'more options' on a cohort or action and then be brought to an entirely different context. And the user can potentially loose the insight they're working on. I wonder if that should be a "new tab" link with a clear icon?

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 commented:
> * Should the owner be visible in the pop-over? Feels a bit odd to add an owner inline, but not be able to see it in the view mode. If we don't think it's valuable, maybe we just remove it from the pop-up?
> * I think it can be a bit jarring to click 'more options' on a cohort or action and then be brought to an entirely different context. And the user can potentially loose the insight they're working on. I wonder if that should be a "new tab" link with a clear icon?

Had a quick call with Rick to address these:
1. Ownership doesn't make sense yet as it holds no weight in how definitions are edited. Opting to remove from these from edit mode.
2. New tab icon seems like the way to go!

alexkim205 committed:
address feedback

alexkim205 committed:
popup in new tab

alexkim205 committed:
fix ts error

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
view fix and tags styling

mariusandra commented on frontend/src/lib/components/DefinitionPopup/definitionPopupLogic.ts:
```diff
@@ -24,8 +43,173 @@ export const definitionPopupLogic = kea<definitionPopupLogicType<DefinitionPopup
+            {
+                setDefinition: ({ item }) => item as TaxonomicDefinitionTypes,
+                handleSave: async (_, breakpoint) => {
+                    await breakpoint(100)
```
Why wait?

alexkim205 replied:
true true, the button will be disabled anyways when this is triggered

mariusandra commented on frontend/src/lib/hooks/useOutsideClickHandler.ts:
```diff
@@ -22,8 +22,8 @@ export function useOutsideClickHandler(
             // Only attach event listeners if there's something to track
-            document.addEventListener('click', handleClick)
-            return () => document.removeEventListener('click', handleClick)
+            document.addEventListener('mouseup', handleClick)
+            return () => document.removeEventListener('mouseup', handleClick)
```
`click` also fires on touch events. What implications does this change have for e.g. iPad users?

alexkim205 replied:
good catch. I think it's correct to listen up "end" events in general so we probably want to also listen to touchend events to support mobile users.

mariusandra commented on frontend/src/types.ts:
```diff
@@ -228,6 +228,7 @@ export interface ActionType {
     slack_message_format?: string
     steps?: ActionStepType[]
     created_by: UserBasicType | null
+    tags?: string[] // TODO: to be added with global tags
```
`// merge conflict` ?

mariusandra commented on frontend/src/lib/components/TaxonomicFilter/InfiniteList.tsx:
```diff
@@ -215,6 +221,7 @@ const selectedItemHasPopup = (
                 TaxonomicFilterGroupType.Events,
                 TaxonomicFilterGroupType.CustomEvents,
                 TaxonomicFilterGroupType.EventProperties,
+                TaxonomicFilterGroupType.NumericalEventProperties,
```
Thanks for the fixes here!

mariusandra reviewed (dismissed):
This is sleek! Found some small things to comment on.

UX wise, love the fade effect. Just to test its limits, I went to see it in a drawer:

![image](https://user-images.githubusercontent.com/53387/154141788-778582a0-d082-40a8-9137-413f364fbe22.png)

Looked good, but I can't scroll to see the save button. Didn't check if it works with modals either.

alexkim205 commented on frontend/src/lib/components/DefinitionPopup/definitionPopupLogic.ts:
```diff
@@ -24,8 +43,173 @@ export const definitionPopupLogic = kea<definitionPopupLogicType<DefinitionPopup
+                            // Action Definitions
+                            const _action = definition as ActionType
+                            definition = await api.update(`api/projects/@current/actions/${_action.id}`, _action)
+                            actionsModel?.isMounted() && actionsModel.actions.updateAction(definition as ActionType)
```
that's pretty sweet!

alexkim205 commented:
Thanks, addressed everything above and fixed a bug with non-license views. It was a nice surprise that it worked on modals without any tweaking.

https://user-images.githubusercontent.com/13460330/154152849-addd996c-806a-4d18-ae28-52fc5a12a141.mov

Hijacked save to dashboard modal to test this, since we don't have insight filters in any modals atm.

https://user-images.githubusercontent.com/13460330/154152866-d07037bd-5715-463b-9a83-a1403f39acc2.mov

alexkim205 committed:
addressed feedback

clarkus reviewed (dismissed):
Looks good to me overall. I noticed that we're not formatting some numeric values in places. Can we include that here? For example the value `123456789` would be formatted as `123,456,789`.

alexkim205 committed:
format numbers

mariusandra reviewed (approved):
Just still the bug in the cohort drawer: 
![2022-02-15 22 52 21](https://user-images.githubusercontent.com/53387/154155731-a3c98244-d020-4ef4-9745-7a2f95841285.gif)

... and did you verify for sure that the same case works in a modal? I mean when the popup is just too big to fit on the screen for whatever reason.

After those, it's good enough to 🚢 , pending changes from merging #8567 probably.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
rerender popper

alexkim205 commented:
Ah, my bad I assumed that was the same as another overflow bug. Here it is working now:

https://user-images.githubusercontent.com/13460330/154160044-2cf3bc70-0da0-4478-92c8-d211a4cbed51.mov

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/edit-definition-popup

alexkim205 committed:
use onchange

alexkim205 committed:
fix tests
